File: services/bff-water-control/src/config/redis.js
const redis = require('redis');
const logger = require('../utils/logger');

class RedisConfig {
  constructor() {
    this.subscriber = null;
    this.publisher = null;
    this.isConnected = false;
    this.reconnectAttempts = 0;
    this.maxReconnectAttempts = 10;
    this.backoffMultiplier = 100;
  }

  async createRedisSubscriber() {
    try {
      // Create subscriber client
      this.subscriber = redis.createClient({
        socket: {
          host: process.env.REDIS_HOST || 'localhost',
          port: parseInt(process.env.REDIS_PORT || '6379'),
          reconnectStrategy: (attempts) => this.reconnectStrategy(attempts)
        },
        password: process.env.REDIS_PASSWORD || undefined
      });

      // Create publisher client
      this.publisher = this.subscriber.duplicate();

      // Set up event handlers for subscriber
      this.subscriber.on('connect', () => {
        this.isConnected = true;
        this.reconnectAttempts = 0;
        logger.info('Redis subscriber connected successfully');
      });

      this.subscriber.on('error', (err) => this.handleRedisError(err));
      this.subscriber.on('end', () => {
        this.isConnected = false;
        logger.warn('Redis subscriber connection closed');
      });

      // Set up event handlers for publisher
      this.publisher.on('connect', () => {
        logger.info('Redis publisher connected successfully');
      });

      this.publisher.on('error', (err) => {
        logger.error('Redis publisher error:', err);
      });

      // Connect both clients
      await Promise.all([
        this.subscriber.connect(),
        this.publisher.connect()
      ]);

      this.isConnected = true;
      return { subscriber: this.subscriber, publisher: this.publisher };

    } catch (error) {
      logger.error('Failed to create Redis clients:', error);
      this.isConnected = false;
      throw error;
    }
  }

  reconnectStrategy(attempts) {
    this.reconnectAttempts = attempts;
    
    if (attempts > this.maxReconnectAttempts) {
      logger.error(`Redis reconnection failed after ${attempts} attempts`);
      return false; // Stop reconnecting
    }

    const delay = Math.min(attempts * this.backoffMultiplier, 3000);
    logger.info(`Redis reconnecting... attempt ${attempts}, delay ${delay}ms`);
    return delay;
  }

  handleRedisError(error) {
    logger.error('Redis error:', error.message);
    
    // Implement circuit breaker pattern
    if (error.code === 'ECONNREFUSED' || error.code === 'ENOTFOUND') {
      logger.warn('Redis server not reachable, operating in degraded mode');
      this.isConnected = false;
    }
  }

  getStatus() {
    return {
      connected: this.isConnected,
      subscriberReady: this.subscriber?.isReady || false,
      publisherReady: this.publisher?.isReady || false,
      reconnectAttempts: this.reconnectAttempts
    };
  }

  async disconnect() {
    try {
      if (this.subscriber) {
        await this.subscriber.quit();
        this.subscriber = null;
      }
      if (this.publisher) {
        await this.publisher.quit();
        this.publisher = null;
      }
      this.isConnected = false;
      logger.info('Redis clients disconnected');
    } catch (error) {
      logger.error('Error disconnecting Redis:', error);
    }
  }
}

module.exports = new RedisConfig();

File: services/bff-water-planning/src/config/redis.py
"""Redis configuration for event publishing."""
import os
import logging
from typing import Optional
import redis.asyncio as redis

logger = logging.getLogger(__name__)


class RedisConfig:
    """Redis configuration and connection manager."""
    
    def __init__(self):
        self.publisher: Optional[redis.Redis] = None
        self.is_connected = False
        self.connection_attempts = 0
        self.max_retries = 10
        
    async def create_redis_client(self) -> Optional[redis.Redis]:
        """Create and configure Redis publisher client."""
        try:
            self.publisher = redis.Redis(
                host=os.getenv('REDIS_HOST', 'localhost'),
                port=int(os.getenv('REDIS_PORT', '6379')),
                password=os.getenv('REDIS_PASSWORD', None),
                decode_responses=True,
                socket_connect_timeout=5,
                socket_keepalive=True,
                socket_keepalive_options={},
                retry_on_timeout=True,
                retry_on_error=[ConnectionError, TimeoutError]
            )
            
            # Test connection
            await self.publisher.ping()
            self.is_connected = True
            logger.info("Redis publisher connected successfully")
            return self.publisher
            
        except Exception as e:
            logger.error(f"Failed to connect to Redis: {e}")
            self.is_connected = False
            return None
    
    def get_redis_status(self) -> dict:
        """Get current Redis connection status."""
        return {
            'connected': self.is_connected,
            'connection_attempts': self.connection_attempts,
            'publisher_ready': self.publisher is not None
        }
    
    async def disconnect(self):
        """Disconnect Redis client."""
        if self.publisher:
            await self.publisher.close()
            self.publisher = None
            self.is_connected = False
            logger.info("Redis publisher disconnected")


# Singleton instance
redis_config = RedisConfig()